augur/curate/apply_geolocation_rules.py

```diff
@@ -0,0 +1,222 @@
+"""
+Applies user curated geolocation rules to the geolocation fields.
+"""
+from collections import defaultdict
+from augur.errors import AugurError
+from augur.io.print import print_err
+from augur.utils import first_line
+
+
+class CyclicGeolocationRulesError(AugurError):
+    pass
+
+
+def load_geolocation_rules(geolocation_rules_file):
```

Comment:
silly question #2: do we eschew types in `augur`?

Reply:
We don't, you'll find some here and there in two forms:

1. Types in docstrings

    https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/augur/io/metadata.py#L33-L39

2. Type annotations

    https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/augur/filter/subsample.py#L383-L386

These are picked up and rendered in docs such as [`augur.io.read_metadata`](https://docs.nextstrain.org/projects/augur/en/stable/api/public/augur.io.html#augur.io.read_metadata):

<img width="714" alt="augur.io.read_metadata" src="https://github.com/nextstrain/augur/assets/13424970/cc51cd7d-5671-479e-843a-7c634f713051">

But typing isn't ubiquitous and what's typed [isn't even validated](https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/pyrightconfig.json#L16).

Historically most of the codebase has been untyped and we as a team have only recently begun using types in untyped languages. It hasn't been a priority to go back and address, but that's something I would love to see happen.

Reply:
> most of the codebase has been untyped and we as a team have only recently begun using types in untyped languages. It hasn't been a priority to go back and add more types to existing code, but that's something I would love to see happen.

one way to expand the use of types would be to have an general expectation that any new code being added would (at a minimum) annotate method signatures...